pkg/packages/pod.go

```diff
@@ -91,16 +93,20 @@ func GetSpecContainer(spec v1.PodSpec, name string, initContainer bool) (v1.Cont
 	return image, nil
 }
 
-// GetMatchingContainer returns the container from the given set of containers that matches the
-// given name.  If the given container list has only 1 item then the name field is ignored and
-// that container is returned.
+// GetMatchingContainer returns the container from the given set of containers
+// that matches the given name. If the given container list has only 1 item and
+// no name is specified then that container is returned. If more than one
+// container exists and no name is specified it attempts to access a name from
+// env vars, before then checking for the existence of the container.
 func GetMatchingContainer(containers []v1.Container, name string) (v1.Container, error) {
 	var result *v1.Container
-	if len(containers) == 1 {
-		// if there is only one pod, use its image rather than require a set container name
+	if len(containers) == 1 && name == "" {
+		// if there is only one container and no name is specified, return it
 		result = &containers[0]
 	} else {
-		// if there are multiple pods, we require the container to have the expected name
+		if name == "" {
+			name = os.Getenv(ContainerNameEnvVar)
```

Comment:
`GetMatchingContainer` function already takes `name` as an argument, wouldn't it be cleaner if we leave setting the name of the container properly to the caller instead? Tracking the name parameter, I found it is set as empty here:  https://github.com/crossplane/crossplane/blob/0d1aeec8b505bf80de46f16bf25aaed5921a3cfa/pkg/packages/executorinfo.go#L87

Reply:
@turkenh I totally agree. However, this package is already in a bit of a weird pattern IMO because it uses environment variables for `GetRunningPod` so I was attempting to mirror that here. I would be fine going either way. It think some of this code should be replaced soon anyway.

Reply:
The `PACKAGE_MANAGER_IMAGE` is used in debugging contexts, where the image name can not be determined (there is no running pod, just a process).  The image name is needed so `manage` can create `unpack` jobs.

One other scenario where this is used is: 

https://github.com/operator-framework/community-operators/blob/master/community-operators/crossplane/0.11.0/crossplane.v0.11.0.clusterserviceversion.yaml#L403-L404

In this scenario, the package manager is run with an OLM side-car, and the `manage` process could not determine which pod container represented Crossplane's package manager.

Reply:
@displague thanks for that context! Sounds like we could just set https://github.com/crossplane/crossplane/pull/1660/files#diff-5024fc0cbab4a78ee75f7db9c2d696dcR76 to `PACKAGE_MANAGER_IMAGE` as in the OLM case and then pass that from the caller of this function. Does that sound good to you?

Reply:
My concern wasn't about this PR but how external uses could benefit from it.

I think the chart name is fine with how you are matching it now, since it is the same here: https://github.com/crossplane/crossplane/pull/1660/files#diff-5024fc0cbab4a78ee75f7db9c2d696dcR63.

In the OLM manifest, you could replace the `PACKAGE_MANAGER_IMAGE` definition with `CONTAINER_NAME` (`crossplane`) at:
https://github.com/operator-framework/community-operators/blob/master/community-operators/crossplane/0.11.0/crossplane.v0.11.0.clusterserviceversion.yaml#L403-L404

Reply:
@displague I _think_ the changes as proposed here would not actually affect the OLM case at all. If we keep `PACKAGE_MANAGER_IMAGE` as is then `GetMatchingContainer` is never called. Therefore, we would have the following cases:
1. If `PACKAGE_MANAGER_IMAGE` is supplied then we will always use it (useful for overriding in debugging as you mentioned, or specifying a different unpack image than package manager image).
2. If `CONTAINER_NAME` is supplied then we will use it to find the container and it's corresponding image (safer than not supplying in the case that a sidecar is injected).
3. If neither are supplied, if there is only one container then we will use it. Otherwise we will fail.

Reply:
It's one less line to tweak in the OLM manifest if that manifest adopts the new `CONTAINER_NAME` option.